File: ios/Sources/RynBridge/CallbackRegistry.swift
import Foundation

public actor CallbackRegistry {
    private var callbacks: [String: CheckedContinuation<BridgeResponse, Error>] = [:]
    private var timeoutTasks: [String: Task<Void, Never>] = [:]

    public init() {}

    public func register(id: String, timeout: TimeInterval) async throws -> BridgeResponse {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<BridgeResponse, Error>) in
            Task { self.store(id: id, continuation: continuation, timeout: timeout) }
        }
    }

    private func store(id: String, continuation: CheckedContinuation<BridgeResponse, Error>, timeout: TimeInterval) {
        callbacks[id] = continuation

        let timeoutTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
            guard !Task.isCancelled else { return }
            await self?.handleTimeout(id: id)
        }
        timeoutTasks[id] = timeoutTask
    }

    private func handleTimeout(id: String) {
        guard let continuation = callbacks.removeValue(forKey: id) else { return }
        timeoutTasks.removeValue(forKey: id)?.cancel()
        continuation.resume(throwing: RynBridgeError(code: .timeout, message: "Request \(id) timed out"))
    }

    public func resolve(id: String, response: BridgeResponse) -> Bool {
        guard let continuation = callbacks.removeValue(forKey: id) else { return false }
        timeoutTasks.removeValue(forKey: id)?.cancel()
        continuation.resume(returning: response)
        return true
    }

    public func clear() {
        for (id, continuation) in callbacks {
            continuation.resume(throwing: RynBridgeError(code: .transportError, message: "Bridge disposed, request \(id) cancelled"))
        }
        callbacks.removeAll()
        for (_, task) in timeoutTasks {
            task.cancel()
        }
        timeoutTasks.removeAll()
    }

    public var pendingCount: Int {
        callbacks.count
    }
}

File: android/core/src/main/kotlin/io/rynbridge/core/CallbackRegistry.kt
package io.rynbridge.core

import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class CallbackRegistry {

    private val mutex = Mutex()
    private val callbacks = mutableMapOf<String, CancellableContinuation<BridgeResponse>>()
    private val timeoutJobs = mutableMapOf<String, Job>()

    suspend fun register(id: String, timeout: Long, scope: CoroutineScope): BridgeResponse {
        return suspendCancellableCoroutine { continuation ->
            scope.launch {
                mutex.withLock {
                    callbacks[id] = continuation
                }

                val timeoutJob = scope.launch {
                    delay(timeout)
                    handleTimeout(id)
                }

                mutex.withLock {
                    timeoutJobs[id] = timeoutJob
                }

                continuation.invokeOnCancellation {
                    scope.launch {
                        mutex.withLock {
                            callbacks.remove(id)
                            timeoutJobs.remove(id)?.cancel()
                        }
                    }
                }
            }
        }
    }

    private suspend fun handleTimeout(id: String) {
        mutex.withLock {
            val continuation = callbacks.remove(id) ?: return
            timeoutJobs.remove(id)?.cancel()
            continuation.resumeWithException(
                RynBridgeError(code = ErrorCode.TIMEOUT, message = "Request $id timed out")
            )
        }
    }

    suspend fun resolve(id: String, response: BridgeResponse): Boolean {
        mutex.withLock {
            val continuation = callbacks.remove(id) ?: return false
            timeoutJobs.remove(id)?.cancel()
            continuation.resume(response)
            return true
        }
    }

    suspend fun clear() {
        mutex.withLock {
            for ((id, continuation) in callbacks) {
                continuation.resumeWithException(
                    RynBridgeError(
                        code = ErrorCode.TRANSPORT_ERROR,
                        message = "Bridge disposed, request $id cancelled"
                    )
                )
            }
            callbacks.clear()
            for ((_, job) in timeoutJobs) {
                job.cancel()
            }
            timeoutJobs.clear()
        }
    }

    suspend fun pendingCount(): Int {
        mutex.withLock {
            return callbacks.size
        }
    }
}
